examples/prom/de_parametric_maxwell_greedy.cpp

```diff
@@ -0,0 +1,956 @@
+//
+// Larger example:
+// build_database phase: mpirun -np 8 de_parametric_maxwell_greedy -build_database -greedy-param-min 0.5 -greedy-param-max 1.5 -greedy-param-size 15 -greedysubsize 4 -greedyconvsize 6 -greedyrelerrortol 0.01
+// FOM phase:            mpirun -np 8 de_parametric_maxwell_greedy -visit -fom -f X.XX (create a new solution to compare with, set X.XX to desired frequency)
+// DE phase:             mpirun -np 8 de_parametric_maxwell_greedy  -f X.XX -visit -de -de_f 0.9 -de_cr 0.9 -de_ps 50 -de_min_iter 10 -de_max_iter 100 -de_ct 0.001 -de_min_freq 0.5 -de_max_freq 1.5 (Run interpolative differential evolution to see if target FOM can be matched)
+//
+// Results with frequency 0.8 - Best minimal cost: 0.000367194        Best agent: 0.800017
+
+// This example runs in parallel with MPI, by using the same number of MPI ranks
+// in all phases (build_database, fom, de).
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+#include "linalg/Vector.h"
+#include "linalg/BasisGenerator.h"
+#include "linalg/BasisReader.h"
+#include "mfem/Utilities.hpp"
+#include <cmath>
+#include <set>
+#include <vector>
+#include "algo/greedy/GreedyRandomSampler.h"
+#include "algo/DifferentialEvolution.h"
+
+#ifndef _WIN32
+#include <sys/stat.h>  // mkdir
+#else
+#include <direct.h>    // _mkdir
+#define mkdir(dir, mode) _mkdir(dir)
+#endif
+
+using namespace std;
+using namespace mfem;
+
+// Exact solution, E, and r.h.s., f. See below for implementation.
+void E_exact(const Vector &, Vector &);
+void f_exact(const Vector &, Vector &);
+double freq = 1.0, kappa;
+int dim;
+
+int myid;
+int num_procs;
+
+// 2. Parse command-line options.
+const char *mesh_file = "../data/star.mesh";
+int order = 1;
+bool static_cond = false;
+bool pa = false;
+const char *device_config = "cpu";
+bool visualization = false;
+bool visit = false;
+bool fom = false;
+bool offline = false;
+bool online = false;
+int precision = 8;
+double coef = 1.0;
+
+// Greedy algorithm variables
+bool build_database = false;
+bool use_database = false;
+double greedy_param_space_min = 1.0;
+double greedy_param_space_max = 1.2;
+int greedy_param_space_size = 5;
+double greedy_relative_error_tol = 0.01;
+int greedy_subset_size = 2;
+int greedy_convergence_subset_size = 3;
+double local_rom_freq = 0.0;
+double curr_freq = 0.0;
+vector<string> basisIdentifiers;
+CAROM::GreedySampler* greedy_sampler = NULL;
+bool calc_rel_error;
+std::string curr_basis_identifier;
+const char *baseoutputname = "";
+bool calc_err_indicator = false;
+
+// DE variables
+bool de = false;
+double de_min_freq= -DBL_MAX;
+double de_max_freq = DBL_MAX;
+double target_freq = freq;
+double de_F = 0.8;
+double de_CR = 0.9;
+int de_PS = 50;
+int de_min_iter = 10;
+int de_max_iter = 100;
+double de_ct = 0.001;
+#ifdef MFEM_USE_AMGX
+bool useAmgX = false;
+#endif
+
+// Define the process for solving Maxwell equation. We will repeatedly call this function later
+double simulation()
```

Comment:
Is this exactly the same process as used in the maxell_local_rom_greedy example? If so, it might be good to pull this out into a separate file eventually, but that doesn't have to be done for this PR.

Reply:
The processes are very similar, but there are some conditions added just to check if we are running DE or not. It might be possible to make separate files just for the solvers with this, and also with a few other examples.